### ISSUE TITLE
Provider-ingested entities not being eager deleted

### ISSUE THREAD

**Issue by GoFightNguyen** [backstage/backstage#12883]
## Expected Behavior

We expect provider-ingested Entities to be eager deleted.

According to the docs
> Entity providers - not processors - are subject to eager deletion of entities...When a provider issues a deletion of an entity in its bucket, that entity as well as the entire tree of entities processed out of it, if any, are considered for immediate deletion. Note "considered" - they are deleted if and only if they would otherwise have become orphaned (no other parent entities emitting them).

## Actual Behavior

We are seeing a mix of behavior.
1. Entities are marked as orphaned but not eager deleted
1. Entities are not being marked as orphaned despite the provider no longer ingesting them (they are removed from the external source)

### DB Behavior

For both behaviors above, we noticed the `refresh_state_references` table have 0 entries for the problem entities (whether as the `source_entity_ref` or `target_entity_ref`).

Note (maybe unrelated): The `refresh_state_references` table has multiple entries for provider-ingested entities we expect to be neither orphaned nor deleted.
Example:
| id | source_key | source_entity_ref | target_entity_ref |
| --- | --- | --- | --- |
| 441616 | LdapOrgEntityProvider:okta-ldap | NULL | user:default/jon-doe |
| 499381 | LdapOrgEntityProvider:okta-ldap | NULL | user:default/jon-doe |

## Steps to Reproduce

We are unable to reproduce this locally.

Locally, we created a custom `EntityProvider` in an attempt to reproduce.
The expected behavior occurred.

## Context

Despite using the `LdapOrgEntityProvider`, we have 500+ orphaned user entities.
In contrast, we have user entities that should also be eager deleted AND are not flagged as orphan.

Also, we have seen the same behavior from a custom `EntityProvider` used in production.

We did not notice this erroneous behavior until after our upgrade to Backstage 1.4.0; not sure if this is a coincidence.
We are positive that some months ago eager deletion of provider-ingested entities was occurring as expected.

## Your Environment

Backstage version: 1.4.0
Postgres version: 13.3

**Comment by benjdlambert** [comment]
Gonna tag this as `after-vacations` and @freben in here too, something for the Catalog SIG perhaps?

**Comment by freben** [comment]
Hey, thanks for the detailed report! In particular, the duplicate rows in the `refresh_state_references` table was a superb thing to spot, because that could hint at some concurrent updates trampling over each other.

We'll have to look closer at this. In the mean time, could I ask how you are scheduling your entity provider in your backend code? Are you using `env.scheduler.scheduleTask()` to make sure that only one host runs the update at any given time? Of course, even if you didn't, we shouldn't let the database end up in an unexpected state - but it would help us understand the nature of how this happens.

**Comment by GoFightNguyen** [comment]
The `LdapOrgEntityProvider` (and our custom provider which ingests a different type of entity) are scheduled using `env.scheduler.createScheduledTaskRunner()`

```typescript

```

**Comment by GoFightNguyen** [comment]
@freben Any advice on what to do regarding the "Actual Behavior" section from the original post?
For 1, should we manually delete/unregister them?
For 2, I'm not sure how to identify what should be orphaned without manually querying the source to see if the entity still exists

**Comment by freben** [comment]
Yes, manual deletion is always an option if you are sure it should indeed be gone, for both cases. Regarding 2, yeah a small cleanup script of sorts might help, but that does sound frustrating. If you have access to the raw database, I guess deleting everything that has zero incoming references (that are never mentioned as a `target_entity_ref` anywhere) should be possible to treat as junk. 🤔

**Comment by GoFightNguyen** [comment]
> We'll have to look closer at this. In the mean time, could I ask how you are scheduling your entity provider in your backend code? Are you using `env.scheduler.scheduleTask()` to make sure that only one host runs the update at any given time? Of course, even if you didn't, we shouldn't let the database end up in an unexpected state - but it would help us understand the nature of how this happens.

@freben There is a chance that the backend has multiple instances running at a time because we use a rolling deploy in k8s. Regardless, we are using `env.scheduler.scheduleTask()`

**Comment by GoFightNguyen** [comment]
Here are the DB commands we ran

### Identify and Delete entities that are not a `target_entity_ref`
```sql
select * from refresh_state
	left join refresh_state_references
	on refresh_state.entity_ref = refresh_state_references.target_entity_ref
where refresh_state_references.id is null;
```
```sql
delete from refresh_state
	where entity_id in (
		select entity_id from refresh_state
		left join refresh_state_references
		on refresh_state.entity_ref = refresh_state_references.target_entity_ref
		where refresh_state_references.id is null
	);
```

### Delete duplicates in `refresh_state_references`
```sql
DELETE FROM refresh_state_references
	where id not in (
		select min(id) from refresh_state_references 
			group by target_entity_ref
		);
```

**Comment by webark** [comment]
@GoFightNguyen How did those DB commands work for you? Did you have any long term issues with those?

**Comment by webark** [comment]
ok. cause @GoFightNguyen I just discovered that we have 46 million records in the `refresh_state_references` table in our staging instance.

**Comment by webark** [comment]
actually.. Maybe it is. I tried the call in the staging DB, and it looks like I will have to batch the deletes the `refresh_state_references` if I don't want to take down my DB.

**Comment by GoFightNguyen** [comment]
Woah, that seems unreal.
fwiw, my `refresh_state_references` table has only 2,921 records and `refresh_state` has 2,738 entities

**Comment by webark** [comment]
There's a good chunk of these that don't have a location key due to us wanting to allow for them to be overwritten when a catalog file if registered. So I wonder if the lack of location key is causing part of the issue. We might do some modifications like increasing the interval between syncs of those records or something.

**Comment by Rugvip** [comment]
Closing this as #15146 and further fix in #16305 should hopefully solve the overall issue here. If anyone encounters similar issues when running Backstage 1.11 or later, please open a new issue.
